clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractHeartbeatRequestManager.java

```diff
@@ -275,98 +242,98 @@ private NetworkClientDelegate.UnsentRequest makeHeartbeatRequest(final long curr
+                    logger.error("{} failed because of {}: {}", heartbeatRequestName(), error, response);
             } else {
-                logger.error("GroupHeartbeat failed because of unexpected exception.", exception);
+                logger.error("{} failed because of unexpected exception.", heartbeatRequestName(), exception);
             }
         });
     }
 
     private void onFailure(final Throwable exception, final long responseTimeMs) {
         this.heartbeatRequestState.onFailedAttempt(responseTimeMs);
-        this.heartbeatState.reset();
-        membershipManager.onHeartbeatFailure(exception instanceof RetriableException);
+        resetHeartbeatState();
+        membershipManager().onHeartbeatFailure(exception instanceof RetriableException);
         if (exception instanceof RetriableException) {
-            String message = String.format("GroupHeartbeatRequest failed because of the retriable exception. " +
-                    "Will retry in %s ms: %s",
-                heartbeatRequestState.remainingBackoffMs(responseTimeMs),
-                exception.getMessage());
+            String message = String.format("%s failed because of the retriable exception. Will retry in %s ms: %s",
+                    heartbeatRequestName(),
+                    heartbeatRequestState.remainingBackoffMs(responseTimeMs),
+                    exception.getMessage());
             logger.debug(message);
         } else {
-            logger.error("GroupHeartbeatRequest failed due to fatal error: " + exception.getMessage());
+            logger.error("{} failed due to fatal error: {}", heartbeatRequestName(), exception.getMessage());
             handleFatalFailure(exception);
         }
     }
 
-    private void onResponse(final ConsumerGroupHeartbeatResponse response, long currentTimeMs) {
-        if (Errors.forCode(response.data().errorCode()) == Errors.NONE) {
-            heartbeatRequestState.updateHeartbeatIntervalMs(response.data().heartbeatIntervalMs());
+    private void onResponse(final HBR response, long currentTimeMs) {
+        if (Errors.forCode(errorCodeForResponse(response)) == Errors.NONE) {
+            heartbeatRequestState.updateHeartbeatIntervalMs(heartbeatIntervalForResponse(response));
             heartbeatRequestState.onSuccessfulAttempt(currentTimeMs);
-            membershipManager.onHeartbeatSuccess(response.data());
+            membershipManager().onHeartbeatSuccess(responseData(response));
             return;
         }
         onErrorResponse(response, currentTimeMs);
     }
 
-    private void onErrorResponse(final ConsumerGroupHeartbeatResponse response,
+    private void onErrorResponse(final HBR response,
                                  final long currentTimeMs) {
-        Errors error = Errors.forCode(response.data().errorCode());
-        String errorMessage = response.data().errorMessage();
+        Errors error = Errors.forCode(errorCodeForResponse(response));
```

Comment:
seems we always end up wrapping the result of `errorCodeForResponse` with Errors.forCode? If so we could simplify maybe by having errorCodeForResponse return an Error already and then just `Errors error = errorCodeForResponse(response);`

Reply:
Agreed.